native_client/java/libdeepspeech/src/main/java/org/mozilla/deepspeech/recognition/DeepSpeechModel.java

```diff
@@ -0,0 +1,151 @@
+        if (ret == null) throw new NullPointerException();
+        return ret;
+    }
+
+    /**
+     * Performs a text to speech call on the recognition and returns a more detailed STT result with additional meta data - not just a sting
+     *
+     * @param audioBuffer the audio buffer storing the audio data in samples / frames to perform the recognition on
+     * @param numSamples  the number of samples / frames in the buffer
+     * @param sampleRate  the amount of samples representing a given duration of audio. sampleRate = Δ samples / Δ time
+     * @return the meta data of transcription
+     * @see SpeechRecognitionResult
+     */
+    @NotNull
+    public SpeechRecognitionResult doSpeechRecognitionWithMeta(@NativeType("const short *") @NotNull ByteBuffer audioBuffer, long numSamples, long sampleRate) {
+        long metaPointer = speechToTextWithMetadata(this.pointer, audioBuffer, numSamples, sampleRate);
+        if (metaPointer == NULL) throw new NullPointerException();
+        return new SpeechRecognitionResult(metaPointer); // Meta pointer is freed as Recognition Result instantly disposes it after copying the values.
+    }
+
+    /**
+     * Allocates a new native recognition structure and returns the pointer pointing to the dynamically allocated memory
+     *
+     * @see DeepSpeechModel#DeepSpeechModel(File, long, long, File, long)
+     */
+    private static long newModel(@NotNull File modelFile, long numCep, long context, @NotNull File alphabetConfigFile, long beamWidth) throws FileNotFoundException {
+        ByteBuffer ptr = ByteBuffer.allocateDirect(NATIVE_POINTER_SIZE).order(ByteOrder.LITTLE_ENDIAN);
+        if (createModel(checkExists(modelFile).getPath(), numCep, context, checkExists(alphabetConfigFile).getPath(), beamWidth, ptr) != 0)
+            throw new RuntimeException("Failed to create recognition!");
+        return getNativePointer(getBufferAddress(ptr));
+    }
+
+    /**
+     * De-allocates the struct memory when this object is garbage collected
+     */
+    @Override
+    protected void deallocateStruct(long pointer) {
+        destroyModel(pointer);
+    }
+
+    /**
+     * @return the pointer to the native struct below
+     */
+    public long getPointer() {
+        return this.pointer;
+    }
+
+    /**
+     * Unsafe function! Consider using {@link #doSpeechToText(ByteBuffer, long, long)} instead.
+     * Performs a text to speech call on the recognition.
+     *
+     * @param audioBufferPointer the audio buffer storing the audio data in samples / frames to perform the recognition on
+     * @param numSamples         the number of samples / frames in the buffer
+     * @param sampleRate         the amount of samples representing a given duration of audio. sampleRate = Δ samples / Δ time
+     * @return the transcription string
+     */
+    @NotNull
+    public String doSpeechToTextUnsafe(@NativeType("const short *") long audioBufferPointer, long numSamples, long sampleRate) {
```

Comment:
Why the Unsafe ?

Reply:
Because it assumes that the memory that the pointer points to is at least numSamples * 2 bytes.

Reply:
But why have a safe and unsafe variant ?

Reply:
You can allocate native heap memory using the Unsafe Java utility. This memory could also be passed to this function. It is recommended to use a ByteBuffer, but this object has limited capabilities depending on what you want to achieve. It has proven to be unsuitable for the SpeechRecognitionAudioStream because the memory of them is also pooled internally in a very strange way after the wrapper object is garbage collected. A stream must be able to quickly receive data and not introduce hundreds of objects when written to. The memory footprint of the application would explode.

Reply:
> A stream must be able to quickly receive data and not introduce hundreds of objects when written to.

Sure, but so far this is not the kind of issues that are really impacting performances, as of now.


> The memory footprint of the application would explode.

We did check that and there was no issue with memory footprint (we did find issues, but not here).

Reply:
I'm not talking about the current bindings, but my SpeechRecognitionAudioStream which does not use the native StreamingState class. It uses a native memory buffer that grows when more capacity is needed. If I just created a new ByteBuffer with the needed capacity and let that old ByteBuffer be garbage collected, that would not be a very good idea.